decimal.go

```diff
@@ -773,6 +774,17 @@ func (d *Decimal) Scan(src interface{}) error {
 	}
 }
 
+// UnmarshalText implements the encoding.TextUnmarshaler interface.
+func (d *Decimal) UnmarshalText(b []byte) error {
+	_, _, err := d.SetString(string(b))
+	return err
+}
+
+// MarshalText implements the encoding.TextMarshaler interface.
+func (d *Decimal) MarshalText() ([]byte, error) {
+	return []byte(d.String()), nil
```

Comment:
This needs to check for d == nil. See: https://github.com/golang/go/blob/c195deb48cf09970c6bb77fad96926e0af102dcf/src/math/big/intmarsh.go#L47

Reply:
What's the expected behavior of `d == nil`?
I was thinking about returning  a nil as `[]byte` and an error but not sure if you had other thoughts. Any preferences on the error message if we go that route?

Reply:
We should do the same as in that link:

```
	if x == nil {
		return []byte("<nil>"), nil
	}
```

Reply:
We'll have the same problem of determining the correct generic marshaled text value for null decimals. Maybe we should only implement the JSON marshal/unmarshal interfaces instead of the generic text ones.

Reply:
Ok I'm changing my mind. I now think we should implement both the text marshaling interfaces on both types, and not the json marshaling interfaces. The marshaling funcs will return `"<nil>"` if d is nil. NullDecimal will return `"null"` if not valid. Both will return the string value otherwise. The unmarshling funcs don't try to detect `"<nil>"` (same as math/big.Int), but instead call just SetString. However the NullDecimal unmarshal func first checks if the string is `"null"`, and if so, sets the value to null. Can you see any problems with this?

Reply:
Your proposal works for me and my use case (I'm only working with JSON for now).

I understand the value of implementing the text marshaling interfaces instead of the json ones. It's nice to have a single serialization for all text formats (JSON, XML, CSV)

However I'm concerned that there's no clear generic null value. When marshaling to XML or CSV, users may want to use values other than `"null"` for interoperability with other (probably existing) systems e.g. `""`, `"nil"`, `"N/A"`, etc
We could document the behavior you proposed above to correctly set expectations with users and/or make the marshaled `nil` Decimal value configurable.

I'm fine w/ implementing the text marshaling interfaces for now and adding the appropriate docs. We can make the text marshaled `nil` value configurable later

Reply:
Yes, I agree that's a concern. I'm still going back and forth about this, but essentially that's why I proposed just not implementing anything for the NullDecimal type, which was added just for SQL compat. That would force users to encode nulls how they needed to.

Reply:
I see value in `NullDecimal` supporting json encoding.
One use case is if you're consuming an external json API with null decimal values, it's nice to be able to unmarshal the json response directly to a `NullDecimal`.

Do you recommendation unmarshaling into a `*Decimal` instead?
If so, we can drop text marshaling/unmarshaling support for `NullDecimal` and have `Decimal.MarshalText()` return `<nil>` when `*d` is `nil`

Reply:
My preference is to add these methods only to the Decimal type and leave NullDecimal alone for now. Adding only the Marshal method without a corresponding Unmarshal method on NullDecimal feels like a bad design, and so I think it should be skipped until we have better reasons for making a correct choice.

Reply:
I've pushed changes addresses all outstanding issues that I'm aware of. Thanks again for being open minded and collaborative!